nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -484,83 +615,152 @@ impl DataStore {
+            .set((
+                dsl::migration_id.eq(Some(migration_id)),
+                dsl::target_propolis_id.eq(Some(target_propolis_id)),
+                // advance the generation
+                dsl::state_generation.eq(dsl::state_generation + 1),
+                dsl::time_state_updated.eq(Utc::now()),
+            ))
+            .returning(Instance::as_returning())
+            .get_result_async::<Instance>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
             .await
-            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
-
-        let instance_updated = match result.instance_status {
-            Some(UpdateStatus::Updated) => true,
-            Some(UpdateStatus::NotUpdatedButExists) => false,
-            None => false,
-        };
+            .map_err(|error| {
+                Error::conflict(format!(
+                    "cannot set migration ID {migration_id} for instance \
+                     {instance_id} (perhaps a previous migration is already \
+                     set): {error:#}"
+                ))
+            })
+    }
 
-        let vmm_updated = match result.vmm_status {
-            Some(UpdateStatus::Updated) => true,
-            Some(UpdateStatus::NotUpdatedButExists) => false,
-            None => false,
-        };
+    /// Unsets the migration IDs set by
+    /// [`DataStore::instance_set_migration_ids`].
+    ///
+    /// This method will only unset the instance's migration IDs if they match
+    /// the provided ones.
+    /// # Returns
+    ///
+    /// - `Ok(true)` if the migration IDs were unset,
+    /// - `Ok(false)` if the instance IDs have *already* been unset (this method
+    ///   is idempotent)
+    /// - `Err` if the database query returned an error.
+    pub async fn instance_unset_migration_ids(
+        &self,
+        opctx: &OpContext,
+        instance_id: InstanceUuid,
+        migration_id: Uuid,
+        target_propolis_id: PropolisUuid,
+    ) -> Result<bool, Error> {
+        use db::schema::instance::dsl;
 
-        let migration_updated = if migration.is_some() {
-            Some(match result.migration_status {
-                Some(UpdateStatus::Updated) => true,
-                Some(UpdateStatus::NotUpdatedButExists) => false,
-                None => false,
+        let instance_id = instance_id.into_untyped_uuid();
+        let target_propolis_id = target_propolis_id.into_untyped_uuid();
+        let updated = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            .filter(dsl::migration_id.eq(migration_id))
+            .filter(dsl::target_propolis_id.eq(target_propolis_id))
+            .set((
+                dsl::migration_id.eq(None::<Uuid>),
+                dsl::target_propolis_id.eq(None::<Uuid>),
+                // advance the generation
+                dsl::state_generation.eq(dsl::state_generation + 1),
```

Comment:
I think we might still have a little refining to do here. The idea behind `SagaUnwound` was to make it so that during a start/migrate unwind, the unwinding saga would only have to touch the VMM that it was unwinding and not the Instance. That prevents this sort of race:

- migration saga starts, sets migration ID at instance generation 5
- instance updater moves instance to generation 6
- migration saga unwinds, wants to clear migration IDs, can't because generation 6 was already written

I think it might work out to have `set_migration_ids` check to see if either

- there is no target Propolis ID or migration ID, or
- the target Propolis is SagaUnwound *and* the migration is Failed on both sides (i.e. we went through `migration_mark_failed`)

WDYT?

Reply:
Ah, yeah, that makes sense. In that case, I think we need to change the unwinding code in the migrate saga to not touch the instance record, too!

Reply:
Okay, I've made this change in https://github.com/oxidecomputer/omicron/pull/5749/commits/bbb96471c18cd57d62d53cf315a52ac12b920d8c.

It occurs to me that we might also want to add retry-with-backoff behavior to the `set_migration_ids` action in the `instance-migrate` saga. This is because an unwinding migrate saga will set the VMM to `SagaUnwound` and the migration to `Failed` in separate reverse actions, so there's a window of time where a VMM can be unwound but the migration hasn't been unwound yet, and we can't set new migration IDs. In that case, if we see a `SagaUnwound` VMM, we might want to keep retrying for a bit if the migration has not been failed yet, because that tells us that another saga is in the process of unwinding, and we might be able to succeed if we just waited a tiny bit longer. Of course, we wouldn't want to retry indefinitely, because the other unwinding saga may have gotten stuck.

But, that seemed like something to address separately --- do you think it should happen on this PR or be deferred for a subsequent one?

Reply:
I'd definitely defer this. IIUC, in this case the client should still get a 409 Conflict indicating that a migration is already in progress, and it can always choose to retry once it sees that the existing migration is fully retired.

Reply:
Ah, yeah, good point.